presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java

```diff
@@ -379,11 +380,101 @@ public void testWeightedScheduling()
+        group1.setMaxRunningQueries(2);
+
+        InternalResourceGroup group2 = root.getOrCreateSubGroup("2");
+        group2.setSoftMemoryLimit(new DataSize(1, MEGABYTE));
+        group2.setMaxQueuedQueries(20);
+        group2.setMaxRunningQueries(2);
+        group2.setSchedulingWeight(2);
+        group2.setSchedulingPolicy(QUERY_PRIORITY);
+
+        InternalResourceGroup group1A = group1.getOrCreateSubGroup("a");
+        group1A.setSoftMemoryLimit(new DataSize(1, MEGABYTE));
+        group1A.setMaxQueuedQueries(10);
+        group1A.setMaxRunningQueries(10);
+
+        InternalResourceGroup group1B = group1.getOrCreateSubGroup("b");
+        group1B.setSoftMemoryLimit(new DataSize(1, MEGABYTE));
+        group1B.setMaxQueuedQueries(10);
+        group1B.setMaxRunningQueries(10);
+
+        InternalResourceGroup group2A = group2.getOrCreateSubGroup("a");
+        group2A.setSoftMemoryLimit(new DataSize(1, MEGABYTE));
+        group2A.setMaxQueuedQueries(10);
+        group2A.setMaxRunningQueries(10);
+
+        InternalResourceGroup group2B = group2.getOrCreateSubGroup("b");
+        group2B.setSoftMemoryLimit(new DataSize(1, MEGABYTE));
+        group2B.setMaxQueuedQueries(10);
+        group2B.setMaxRunningQueries(10);
+
+        Set<MockQueryExecution> queries = fillGroupTo(group1A, ImmutableSet.of(), 10, false);
+        queries.addAll(fillGroupTo(group1B, ImmutableSet.of(), 10, false));
+        queries.addAll(fillGroupTo(group2A, ImmutableSet.of(), 10, true));
+        queries.addAll(fillGroupTo(group2B, ImmutableSet.of(), 10, true));
+
+        ResourceGroupInfo info = root.getInfo();
+        assertEquals(info.getNumAggregatedRunningQueries(), 0);
+        assertEquals(info.getNumAggregatedQueuedQueries(), 40);
+
+        root.setMaxRunningQueries(4);
+        root.processQueuedQueries();
+        info = root.getInfo();
+        assertEquals(info.getNumAggregatedRunningQueries(), 4);
+        assertEquals(info.getNumAggregatedQueuedQueries(), 36);
+        for (Iterator<MockQueryExecution> iterator = queries.iterator(); iterator.hasNext(); ) {
+            MockQueryExecution query = iterator.next();
+            if (query.getState() == RUNNING) {
+                query.complete();
+                iterator.remove();
+            }
+        }
+
+        root.processQueuedQueries();
+        info = root.getInfo();
+        assertEquals(info.getNumAggregatedRunningQueries(), 4);
+        assertEquals(info.getNumAggregatedQueuedQueries(), 32);
+
+        root.setMaxRunningQueries(10);
+        root.processQueuedQueries();
+        info = root.getInfo();
+        assertEquals(info.getNumAggregatedRunningQueries(), 4);
+        assertEquals(info.getNumAggregatedQueuedQueries(), 32);
+
+        group2.setMaxRunningQueries(10);
+        root.processQueuedQueries();
+        info = root.getInfo();
+        assertEquals(info.getNumAggregatedRunningQueries(), 10);
+        assertEquals(info.getNumAggregatedQueuedQueries(), 26);
```

Comment:
Might be nice to simplify the math in these tests a bit. It's possible to follow if you try, but you might be able to come up with some simpler numbers.